Add command to move a calendar from an user to another

tcitworld opened the pull request:
Implement #5158

Command is : `occ dav:move-calendar %name% %userorigin% %userdestination% [-f]` where `-f` allows to delete the calendar shares with a group if the destination user is not in this group. Otherwise it just fails.

The `uri` name of the calendar is needed, so maybe it would be nice to suggest names based on the actual display names if none are found at first.

A few tests would be nice (but there's not much in nextcloud to have a look at 😉  )

- [x] Test ListCalendars command

tcitworld committed:
Add command to move a calendar from an user to another

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

Add a basic check for displaynames in case we fail to get calendar from uri and put some sf console styles & refactor a bit

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

basic Tests

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

add forgotten createNamedParameter()

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

nickvergessen commented:
Hmm interesting, this is like the move files command right? or whats the idea behind it?

tcitworld commented:
Yes. See https://github.com/nextcloud/server/issues/5158

georgehrke commented:
> Hmm interesting, this is like the move files command right? or whats the idea behind it?

From the original report:

> User A creates a calendar and shared it with a group B. But as soon as User A leaves the company, the calendar may be lost, because the user is disabled in LDAP and his data will be deleted from Nextcloud.
 As administrator I need any kind of possibility to either change the owner of the calendar to another group member or (maybe better) create a calendar which is assigned to a group. I think for the first step it would suffice if the current owner can set a new owner.

georgehrke commented on apps/dav/lib/Command/MoveCalendar.php:
```diff
@@ -0,0 +1,162 @@
+		$calendar = $this->caldav->getCalendarByUri(self::URI_USERS . $userOrigin, $name);
+
+		if (null === $calendar) {
+			/**  If we got no matching calendar with URI, let's try the display names */
```
I'm not really fond of this behavior, especially as we don't implement the calendar standard properly here and allow a user to have multiple calendars with the same displayname.

If an admin - we are talking about cli access here - is unable to provide us with a proper calendar uri, we should simply show an error message rather then starting guessing.

nickvergessen replied:
Maybe list all the calendar uris instead?

MorrisJobke commented:
@georgehrke @nickvergessen @tcitworld Could we move this forward? Would be sad to have this lying around even longer.

georgehrke commented:
@tcitworld Do you mind if I take over?

tcitworld commented:
Help yourself ;)

georgehrke committed:
add list-calendars command

Signed-off-by: Georg Ehrke <developer@georgehrke.com>

Rebase and delete URI suggestion part

Invite to use occ dav:list-calendars instead

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

Fix autoload

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

Use injection & test everything

And rebase

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

Add test for ListCalendars and refactoring

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

Fix indentation

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

MorrisJobke commented:
Nothing for 14 it seems -> moved to 15

tcitworld commented:
Wonder why `checkers` CI failed. :thinking:

danielkesselberg commented:
> Wonder why `checkers` CI failed. thinking

![image](https://user-images.githubusercontent.com/3902676/46571628-92be3e80-c978-11e8-8c82-985584549119.png)

tcitworld commented:
Thanks, not easy to spot !

blizzz commented on apps/dav/lib/Command/ListCalendars.php:
```diff
@@ -0,0 +1,120 @@
+		$principalBackend = new Principal(
+			$this->userManager,
+			$this->groupManager,
+			\OC::$server->getShareManager(),
```
all those \OC::$server->getXYS things should be provided in the constructor

blizzz commented on apps/dav/lib/Command/MoveCalendar.php:
```diff
@@ -0,0 +1,162 @@
+		$principalBackend = new Principal(
+			$this->userManager,
+			$this->groupManager,
+			\OC::$server->getShareManager(),
```
also here, all the \OC::$server->getThings should be provided to the constructor

blizzz commented on apps/dav/lib/Command/MoveCalendar.php:
```diff
@@ -0,0 +1,162 @@
+			list($prefix, $group) = str_split($share['href'], 28);
+			if ('principal:principals/groups/' === $prefix && !$this->groupManager->isInGroup($userDestination, $group)) {
+				if ($force) {
+					$this->caldav->updateShares(new Calendar($this->caldav, $calendar, $this->l10n), [], ['href' => 'principal:principals/groups/' . $group]);
```
missing constructor arguments for 'new Calendar`

blizzz commented on apps/dav/tests/unit/Command/MoveCalendarTest.php:
```diff
@@ -0,0 +1,121 @@
+			'userdestination' => 'user2',
+		]);
+	}
+}
```
please add another line to the end

blizzz commented on apps/dav/tests/unit/Command/MoveCalendarTest.php:
```diff
@@ -0,0 +1,121 @@
+			'userorigin' => 'system',
+			'userdestination' => 'user2',
+		]);
+	}
```
Tests for successes, including --force option, are missig

MorrisJobke commented:
I guess we need to move this to 16 😢

tcitworld commented:
Is a test for the ListCalendar command needed too ?

tcitworld commented:
> Is a test for the ListCalendar command needed too ?

Added. :)

blizzz commented on apps/dav/tests/unit/Command/ListCalendarsTest.php:
```diff
@@ -0,0 +1,139 @@
+		$commandTester = new CommandTester($this->command);
+		$commandTester->execute([
+			'user' => self::USERNAME,
+        ]);
```
indentation seems off, see also following lines

tcitworld replied:
> P.S.: also, opening curly brackets should be on the same line. style guidelines

Isn't this obsolete for functions since we have type hints?

blizzz commented:
Sorry for the delay. Would you mind rebasing? I am running into conflicts locally.

tcitworld commented:
Fixed indentation, rebase and cleanup commits.

blizzz commented:
Works nice! 

Just one thing left that I noticed now: it is possible to share to groups you are not member off, unless "Restrict users to only share with users in their groups" is checked in the admin settings. This should be taken into account in checkShares: group shares do not need to be removed unnecessarily.

tcitworld committed:
Don't check group shares if shareWithGroupMembersOnly is false

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

Remove the system user check

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

tcitworld commented:
@blizzz there you go!

blizzz reviewed (approved):
Looks and works well :)

rullzer commented:
Looks good to me. But I'd like @georgehrke to have a look as well.

tcitworld commented on apps/dav/lib/Command/MoveCalendar.php:
```diff
@@ -0,0 +1,167 @@
+		$this->io = new SymfonyStyle($input, $output);
+
+		if (in_array('system', [$userOrigin, $userDestination], true)) {
+			throw new \InvalidArgumentException("User can't be system");
```
At least it warns the sysadmin that tried something clever ?

georgehrke replied:
Well, but it's no warning, it's an exception that you can't circumvent even if you want to :)

Maybe it would be good to instead:
- check if there is a user called system
- If there is, simply move
- If there isn't, show the error message

tcitworld replied:
Removed the whole check.

tcitworld replied:
@georgehrke ^

ChristophWurst commented on apps/dav/lib/Command/ListCalendars.php:
```diff
@@ -0,0 +1,108 @@
+		$this
+			->setName('dav:list-calendars')
+			->setDescription('List all calendars of a user')
+			->addArgument('user',
```
I'd prefer "uid" over "user". That terminology is also used for other commands.

ChristophWurst commented on apps/dav/lib/Command/MoveCalendar.php:
```diff
@@ -0,0 +1,162 @@
+			->addArgument('name',
+				InputArgument::REQUIRED,
+				'Name of the calendar to move')
+			->addArgument('userorigin',
```
"sourceuid"?

ChristophWurst commented on apps/dav/lib/Command/MoveCalendar.php:
```diff
@@ -0,0 +1,162 @@
+			->addArgument('userorigin',
+				InputArgument::REQUIRED,
+				'User who currently owns the calendar')
+			->addArgument('userdestination',
```
"destinationuid"?

ChristophWurst commented on apps/dav/tests/unit/Command/MoveCalendarTest.php:
```diff
@@ -0,0 +1,367 @@
+	/** @var \OCP\IUserManager|\PHPUnit_Framework_MockObject_MockObject $userManager */
+	private $userManager;
+
+	/** @var \OCP\IGroupManager|\PHPUnit_Framework_MockObject_MockObject $groupManager */
```
we updated phpunit a few months ago. the mockobject class is now namespaced (your IDE should help you find the correct class)

ChristophWurst reviewed (dismissed):
Looks great! Added a few suggestions regarding coding style :)

georgehrke commented:
to test:
- [ ] create calendar as user1, share it to user2, move calendar from user1 to user2.

tcitworld committed:
Handle moving calendar to an user who already has the share

Extra:
    * Fix @ChristophWurst style remarks
    * Add a Note that share links have changed when calendars has user shares (see #13603)

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

Fix forgotten test change

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

tcitworld commented:
Added a test for @georgehrke's proposal, calendar won't be moved if it's already shared to the user (add `-f` to do it anyway and remove the share).

Also added a note to tell that shared links will be updated (when there are shares) in the process (see #13603)

tcitworld commented:
Could this be approved ? :slightly_smiling_face:

ChristophWurst commented:
@blizzz is your review still positive? Then we can finally merge this :)

MorrisJobke reviewed (approved):
Tested and works 👍

MorrisJobke commented:
@tcitworld @MariusBluem Mind to add documentation for this command?

tcitworld commented:
@MorrisJobke There you go https://github.com/nextcloud/documentation/pull/1208

tatrapikao commented:
Support of my managed Instance tells me this does not work for accounts with a space in their UID e.g. "john doe" -> is this true and are their any plans possibilities to fix this if true?

tcitworld commented:
@tatrapikao Please open a new issue.

kesselb commented:
Thanks for your message :+1: 

For your next question, please open a new issue. Most people unsubscribe from closed pull requests/issues and may miss your message. 

> "john doe" -> is this true

I cannot confirm.

If your user id contains a space, you may enclose it with quotes. 

For example: 

`occ dav:list-calendars "jane doe"`
`occ dav:move-calendar jane-123 "jane doe" admin`

tatrapikao commented:
Thank you for sorting this out!
